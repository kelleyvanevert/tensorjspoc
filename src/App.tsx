import { Integration } from "./components/Integration";
import { RoomDemoApp } from "./components/RoomDemoApp";

export default function App() {
  return (
    <div className="m-8 flex gap-8">
      {/* <div className="w-1/2 bg-gray-200 min-h-screen">
        <Integration />
      </div> */}
      <div className="w-1/2 bg-gray-200 min-h-screen">
        <RoomDemoApp />
      </div>
    </div>
  );
}
