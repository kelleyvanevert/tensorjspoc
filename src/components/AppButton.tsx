import React from 'react';
import {Text, StyleProp, ViewStyle} from 'react-native';
import {Touchable} from './Touchable';
import {BaseColors} from './colors';
import {StyleSheet} from 'react-native';

type Props = {
  title: string;
  style?: StyleProp<ViewStyle>;
  onPress?: () => void;
};

export function AppButton({title, style, onPress}: Props) {
  return (
    <Touchable onPress={onPress} style={[styles.container, style]}>
      <Text style={styles.title}>{title}</Text>
    </Touchable>
  );
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: BaseColors.green,
    paddingVertical: 12,
    paddingHorizontal: 18,
    borderRadius: 999,
  },
  title: {
    fontFamily: 'Rubik-Bold',
    fontSize: 18,
    color: BaseColors.deepblue,
    textAlign: 'center',
  },
});
