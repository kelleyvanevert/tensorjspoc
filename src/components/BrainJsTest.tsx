import React, {useEffect, useState} from 'react';
// @ts-ignore
import brain from 'brain.js/browser';
import {Text} from 'react-native';
import TrainedModel from './trained_model.json';

export function BrainJsTest() {
  const [output, setOutput] = useState<any>('todo');

  useEffect(() => {
    const net = new brain.NeuralNetwork({
      binaryThresh: 0.5,
      hiddenLayers: [3],
      activation: 'sigmoid',
      leakyReluAlpha: 0.01,
    });

    net.train([
      {input: [0, 0], output: [0]},
      {input: [0, 1], output: [1]},
      {input: [1, 0], output: [1]},
      {input: [1, 1], output: [0]},
    ]);

    // or:
    // net.fromJSON(TrainedModel as any);

    const output = net.run([1, 0]); // [0.9xxx]

    setOutput(output);
  }, []);

  return <Text>{JSON.stringify(output, null, 2)}</Text>;
}
