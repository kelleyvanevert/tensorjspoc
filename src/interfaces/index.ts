export * from './IMoodIndex'
export * from './IExcercise'
export * from './IExerciseFeatures'
export * from './ITrainingData'
export * from './YesNo'
export * from './Exercies'