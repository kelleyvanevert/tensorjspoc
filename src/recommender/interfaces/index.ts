export * from './IContext'
export * from './IExercise'
export * from './IExerciseFeatures'
export * from './ITrainingData'
export * from './YesNo'
export * from './Exercises'
export * from './IRecommendation'
export * from './IRecommendationEngine'
export * from './IEvaluation'
export * from './IOracleState'
export * from './IDemographics'
export * from './INeeds'