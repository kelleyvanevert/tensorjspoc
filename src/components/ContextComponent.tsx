import { useState } from "react";
import { AppButton } from "./AppButton";
import {
  Mood,
  Moods,
  IContext,
  generateContext,
} from "../recommender/interfaces";

type Props = {
  callback: (context: IContext) => void;
};

export function ContextComponent({ callback }: Props) {
  const [moodValues] = useState<Mood[]>(Moods);
  const [currentMood, setCurrentMood] = useState<Mood>(moodValues[0]);

  const onMoodPress = (mood: Mood) => {
    setCurrentMood(mood);
    callback(generateContext(mood));
  };

  return (
    <div>
      <div className="flex flex-wrap gap-2">
        {moodValues.map((mood: Mood) => (
          <AppButton
            title={mood.name}
            key={mood.value}
            onClick={() => onMoodPress(mood)}
          />
        ))}
      </div>
      <div className="mt-2 text-lg">Mood: {currentMood.name}</div>
    </div>
  );
}
