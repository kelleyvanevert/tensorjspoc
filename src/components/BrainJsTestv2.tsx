import React, { useEffect, useState } from 'react';
// @ts-ignore
import brain from 'brain.js/browser';
import { Text } from 'react-native';
import { Student, Activity, Gender, Score, Weather, TimeOfDay, TasteProfile } from '../interfaces'
import { StudentService, getRandomElement, getRandomIntInclusive } from '../services';

export function BrainJsTestv2() {
  const [output, setOutput] = useState<any>('todo');

  const createTrainingDataInput = (iterations: number): { input: Student, output: Score }[] => {
    let result = Array<{ input: Student, output: Score }>();
    const age = getRandomIntInclusive(16, 28);

    const genderValues = Object.keys(Gender)
      .map(n => Number.parseInt(n))
      .filter(n => !Number.isNaN(n))
      .sort((a, b) => a - b);
    const gender = getRandomElement(genderValues);

    const weatherValues = Object.keys(Weather)
      .map(n => Number.parseInt(n))
      .filter(n => !Number.isNaN(n))
      .sort((a, b) => a - b);
    const weather = getRandomElement(weatherValues);

    const timeValues = Object.keys(TimeOfDay)
      .map(n => Number.parseInt(n))
      .filter(n => !Number.isNaN(n))
      .sort((a, b) => a - b);
    const time = getRandomElement(timeValues);

    const weak_knee = getRandomElement([true, false])

    const activityValues = Object.keys(Activity)
      .map(n => Number.parseInt(n))
      .filter(n => !Number.isNaN(n))
      .sort((a, b) => a - b);
    const activity = getRandomElement(activityValues);

    const tasteProfileValues = Object.keys(TasteProfile)
      .map(n => Number.parseInt(n))
      .filter(n => !Number.isNaN(n))
      .sort((a, b) => a - b);
    const taste = getRandomElement(tasteProfileValues);

    for (let index = 0; index < iterations; index++) {
      const student = new StudentService(age, gender, weather, time, weak_knee, activity);
      result.push({
        input: student.createInput(),
        output: student.computeOutput(taste)
      })
    }
    return result;
  }

  useEffect(() => {
    const trainingDataInput = createTrainingDataInput(100);

    const net = new brain.NeuralNetwork({
      hiddenLayers: [4],
      activation: 'sigmoid'
    });
    const stats = net.train(trainingDataInput);
    console.log(stats);

    const testStudent = new StudentService(18, Gender.Male, Weather.Rainy, TimeOfDay.Evening, true, Activity.dancing);
    const output = net.run(testStudent.createInput());

    setOutput(output);
  }, []);

  return <Text>{JSON.stringify(output, null, 2)}</Text>;
}
