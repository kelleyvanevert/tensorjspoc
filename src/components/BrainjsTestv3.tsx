import React, { useEffect, useState } from 'react';
// @ts-ignore
import brain from 'brain.js/browser';
import { Text } from 'react-native';
import { AppButton } from '../components/AppButton';
import { FlatList } from 'react-native-gesture-handler';
import { SafeAreaView } from 'react-native-safe-area-context';
import { TrainingData, IExcercise, IExcerciseFeatures, YesNo, MoodIndex, Exercises } from '../interfaces';

export function BrainJsTestv3() {
    const softmax_beta = 2;
    const exercises: IExcercise[] = Exercises

    const [trainingData, setTrainingData] = useState<IExcerciseFeatures[]>([]);
    const [net, setNeuralNetwork] = useState<brain.NeuralNetwork>(new brain.NeuralNetwork({
        activation: 'sigmoid',
        hiddenLayers: [4]
    }));
    const [mood, setMood] = useState<MoodIndex>(MoodIndex.SoSo);
    const [recommendation, setRecommendation] = useState<IExcercise>();
    const [modelRecommendations, setModelRecommendations] = useState<IExcercise[]>()

    useEffect(() => {
        const trainingData: TrainingData = {
            input: {
                mood_index: MoodIndex.SoSo,
                three_five_mins: YesNo.No,
                five_seven_mins: YesNo.No,
                seven_ten_mins: YesNo.Yes,
                deffuse: YesNo.No,
                zoom_out: YesNo.No,
                feeling_stressed: YesNo.No,
                feeling_angry: YesNo.No,
                mood_boost: YesNo.No,
                self_compassion: YesNo.No
            },
            output: { score: 1 },
        }
        net.train(trainingData);
        compute_recommendation();
        setTrainingData([trainingData.input])
    }, []);

    const compute_recommendation = () => {
        const scores = [];
        const computedReccommendation: IExcercise[] = []
        for (let index = 0; index < exercises.length; index++) {
            const exercise = exercises[index];
            exercise.Value.mood_index = mood;
            exercise.Score = net.run(exercise.Value).score
            computedReccommendation.push(exercise)
            scores.push(exercise.Score || 0)
        }

        setModelRecommendations(computedReccommendation);
        const probs = converToProbabilityDistribution(scores);
        const index = sampleFromProbabilityDistribution(probs);
        setRecommendation(exercises[index]);
    }

    const converToProbabilityDistribution = (scores: number[]): number[] => {
        let probabilities: number[] = []
        let softmax_denomenator = 0
        const numerators: number[] = []
        for (let i = 0; i < scores.length; i++) {
            let softmax_numerator = Math.exp(softmax_beta * scores[i]);
            softmax_denomenator += softmax_numerator
            numerators.push(softmax_numerator);
        }

        for (let i = 0; i < numerators.length; i++) {
            probabilities.push(numerators[i] / softmax_denomenator)
        }
        return probabilities;
    }

    const sampleFromProbabilityDistribution = (probs: number[]): number => {
        const sum = probs.reduce((a, b) => a + b, 0); // [1,2,3,4] = cumulative sum ie. 1+2+3+4=10
        if (sum <= 0) {
            throw Error('probs must sum to a value greater than zero')
        }
        const normalized = probs.map(prob => prob / sum) // [1,2,3,4] = transform ie. [1/10, 2/10, 3/10, 4/10]
        const sample = Math.random()
        let total = 0
        for (let i = 0; i < normalized.length; i++) {
            total += normalized[i]
            if (sample < total) {
                return i
            }
        }
        return -1;
    }

    const btnRating = (rating: number) => {
        let current_exercise: IExcercise | undefined = exercises.find((curr: IExcercise) => curr.InternalName == recommendation?.InternalName)
        if (current_exercise == undefined) {
            throw "Could not find recommendation!"
        }

        if (mood !== undefined) {
            current_exercise.Value.mood_index = mood
        }

        let newData: TrainingData = {
            input: current_exercise.Value,
            output: {
                score: rating
            }
        }

        setTrainingData([...(trainingData || []), newData.input]);
        net.train(newData);
        compute_recommendation();
    }

    const setCurrentMood = (mood: MoodIndex) => setMood(mood)

    return (
        <SafeAreaView>
            <Text>I am feeling {MoodIndex[mood]}</Text>
            <AppButton title="Are you feeling so so?" onPress={() => setCurrentMood(0)}></AppButton>
            <AppButton title="Are you crying?" onPress={() => setCurrentMood(1)}></AppButton>

            <Text>Recommended: {recommendation?.DisplayName}</Text>

            <FlatList data={modelRecommendations || []}
                keyExtractor={(item) => item.InternalName}
                renderItem={({ item }) => <Text>{item.DisplayName} - {item.Score}</Text>}></FlatList>

            <AppButton title="Yes" onPress={() => btnRating(1)}></AppButton>
            <AppButton title="No" onPress={() => btnRating(0)}></AppButton>
        </SafeAreaView >
    );
}