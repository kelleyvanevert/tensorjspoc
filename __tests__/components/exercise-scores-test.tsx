import { fireEvent, render } from '@testing-library/react-native';
import { hasUncaughtExceptionCaptureCallback } from 'process';
import { ExerciseScores } from '../../src/components/ExerciseScores';
import { MoodComponent } from '../../src/components/ContextComponent';
import { IExcercise, Moods } from '../../src/interfaces';

describe("exercise-scores compoment", () => {
    it('should render an empty list of recommendations', () => {
        const tree = render(<ExerciseScores recommendations={[]} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render a recommendation with display name as test and value as 0', () => {
        const recommendation: IExcercise[] = [{
            DisplayName: 'test',
            ClickScore: 0,
            InternalName: '',
            Features: {
                mood_boost: 0,
                mood_value: 0,
                deffuse: 0,
                feeling_angry: 0,
                feeling_stressed: 0,
                five_seven_mins: 0,
                self_compassion: 0,
                seven_ten_mins: 0,
                three_five_mins: 0,
                zoom_out: 0
            }
        }]
        const tree = render(<ExerciseScores recommendations={recommendation} />).toJSON();
        expect(tree).toMatchSnapshot();
    })
})