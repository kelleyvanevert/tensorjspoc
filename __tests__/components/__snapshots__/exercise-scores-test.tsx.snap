// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exercise-scores compoment should render a recommendation with display name as test and value as 0 1`] = `
<RCTScrollView
  collapsable={false}
  data={
    [
      {
        "DisplayName": "test",
        "InternalName": "",
        "Score": 0,
        "Value": {
          "deffuse": 0,
          "feeling_angry": 0,
          "feeling_stressed": 0,
          "five_seven_mins": 0,
          "mood_boost": 0,
          "mood_value": 0,
          "self_compassion": 0,
          "seven_ten_mins": 0,
          "three_five_mins": 0,
          "zoom_out": 0,
        },
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  handlerTag={1}
  handlerType="NativeViewGestureHandler"
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  style={
    {
      "marginTop": 10,
    }
  }
  viewabilityConfigCallbackPairs={[]}
  waitFor={
    [
      {
        "current": null,
      },
      {
        "current": null,
      },
    ]
  }
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "marginBottom": 5,
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#300096",
              "fontFamily": "Rubik",
              "fontSize": 15,
              "lineHeight": 26,
            }
          }
        >
          Name: 
          test
        </Text>
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#300096",
              "fontFamily": "Rubik",
              "fontSize": 15,
              "lineHeight": 26,
            }
          }
        >
          Score: 
          0
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`exercise-scores compoment should render an empty list of recommendations 1`] = `
<RCTScrollView
  collapsable={false}
  data={[]}
  getItem={[Function]}
  getItemCount={[Function]}
  handlerTag={1}
  handlerType="NativeViewGestureHandler"
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  style={
    {
      "marginTop": 10,
    }
  }
  viewabilityConfigCallbackPairs={[]}
  waitFor={
    [
      {
        "current": null,
      },
      {
        "current": null,
      },
    ]
  }
>
  <View />
</RCTScrollView>
`;
